### PR TITLE
Validators

### LINKED ISSUE #995: add requestSignature validator for nextVersion and currentVersion
the request signature should contain a new validator which checks if the `current.version`and `next.version` are not existing in the anchorRepository.

Otherwise it would be possible for a malicious collaborator to get signatures for already used `current.version` and `next.version` identifiers.

### PR DESCRIPTION
**Pull request by vedhavyas**
Closes #995 

Changes:
- Current and Next version Validators while signing the document update
- Anchor Repository Address validator
- Author and Date Validator

### CODE REVIEW

File: documents/bootstrapper.go [hunk 1]
```diff
@@ -50,7 +50,12 @@ func (Bootstrapper) Bootstrap(ctx map[string]interface{}) error {
 		return errors.New("identity service not initialized")
 	}
 
-	ctx[BootstrappedDocumentService] = DefaultService(repo, anchorRepo, registry, didService)
+	cfg, ok := ctx[bootstrap.BootstrappedConfig].(Config)
+	if !ok {
+		return errors.New("documents config not initialised")
```
Comment (vimukthi-git) [inline documents/bootstrapper.go]: Please use typed errors

File: documents/entityrelationship/bootstrapper.go [hunk 2]
```diff
@@ -59,11 +60,16 @@ func (Bootstrapper) Bootstrap(ctx map[string]interface{}) error {
 		return errors.New("identity factory not initialised")
 	}
 
+	anchorRepo, ok := ctx[anchors.BootstrappedAnchorRepo].(anchors.AnchorRepository)
+	if !ok {
+		return errors.New("anchor repository not initialised")
```
Comment (vimukthi-git) [inline documents/entityrelationship/bootstrapper.go]: please use typed errors

File: documents/invoice/bootstrapper.go [hunk 3]
```diff
@@ -51,6 +52,11 @@ func (Bootstrapper) Bootstrap(ctx map[string]interface{}) error {
 		return errors.New("config service not initialised")
 	}
 
+	anchorRepo, ok := ctx[anchors.BootstrappedAnchorRepo].(anchors.AnchorRepository)
+	if !ok {
+		return errors.New("anchor repository not initialised")
```
Comment (vimukthi-git) [inline documents/invoice/bootstrapper.go]: same here

File: documents/purchaseorder/bootstrapper.go [hunk 4]
```diff
@@ -51,14 +52,19 @@ func (Bootstrapper) Bootstrap(ctx map[string]interface{}) error {
 		return errors.New("config service not initialised")
 	}
 
+	anchorRepo, ok := ctx[anchors.BootstrappedAnchorRepo].(anchors.AnchorRepository)
+	if !ok {
+		return errors.New("anchor repository not initialised")
```
Comment (vimukthi-git) [inline documents/purchaseorder/bootstrapper.go]: same here

File: documents/validator.go [hunk 5]
```diff
@@ -359,28 +407,16 @@ func transitionValidator(collaborator identity.DID) Validator {
...
 }
 
-// SignatureRequestValidator returns a validator group with following validators
-// document timestamp for signing validator
-// document author validator
-// base validator
-// signing root validator
-// signatures validator
-// should be used when node receives a document requesting for signature
-func SignatureRequestValidator(sender identity.DID, idService identity.ServiceDID) ValidatorGroup {
```
Comment (vimukthi-git) [inline documents/validator.go]: thanks 👍

File: documents/validator.go [hunk 6]
```diff
@@ -423,8 +459,17 @@ func ReceivedAnchoredDocumentValidator(
...
 // transitionsValidator
 // it should be called when a document is received over the p2p layer before signing
-func RequestDocumentSignatureValidator(idService identity.ServiceDID, collaborator identity.DID) ValidatorGroup {
+func RequestDocumentSignatureValidator(
+	repo anchors.AnchorRepository,
+	idService identity.ServiceDID,
+	collaborator identity.DID,
+	anchorRepoAddress common.Address) ValidatorGroup {
 	return ValidatorGroup{
+		documentTimestampForSigningValidator(),
```
Comment (vimukthi-git) [inline documents/validator.go]: 👍

File: documents/validator.go [hunk 7]
```diff
@@ -328,26 +339,63 @@ func anchoredValidator(repo anchors.AnchorRepository) Validator {
...
+// returns error if the version id is already anchored.
+func versionNotAnchoredValidator(repo anchors.AnchorRepository, id []byte) error {
+	anchorID, err := anchors.ToAnchorID(id)
+	if err != nil {
+		return errors.NewTypedError(ErrDocumentIdentifier, err)
+	}
+
+	_, _, err = repo.GetAnchorData(anchorID)
+	if err == nil {
+		return errors.New("id is already anchored: %v", anchorID.String())
```
Comment (vimukthi-git) [inline documents/validator.go]: same here

File: documents/validator.go [hunk 8]
```diff
@@ -328,26 +339,63 @@ func anchoredValidator(repo anchors.AnchorRepository) Validator {
...
+		return nil
+	})
+}
+
+// anchorRepoAddressValidator validates if the model is using the configured anchor repository address.
+func anchorRepoAddressValidator(anchoredRepoAddr common.Address) Validator {
+	return ValidatorFunc(func(_, model Model) error {
+		addr := model.UsedAnchorRepoAddress()
+		if !bytes.Equal(addr.Bytes(), anchoredRepoAddr.Bytes()) {
+			return errors.New("used anchor address is not the node configured address")
```
Comment (vimukthi-git) [inline documents/validator.go]: same here

**Comment by ved** [comment]
What is it all about?

**Comment by vimukthi-git** [comment]
@ved I commented.

**Comment by ved** [comment]
Why am I looped in?

File: documents/validator_test.go [hunk 9]
```diff
@@ -697,13 +697,13 @@ func TestValidator_anchorRepoAddressValidator(t *testing.T) {
...
-	model.On("UsedAnchorRepoAddress").Return(testingidentity.GenerateRandomDID().ToAddress()).Once()
-	model.On("UsedAnchorRepoAddress").Return(addr).Once()
+	model.On("AnchorRepoAddress").Return(testingidentity.GenerateRandomDID().ToAddress()).Once()
+	model.On("AnchorRepoAddress").Return(addr).Once()
 
 	// failure
 	err := arv.Validate(nil, model)
 	assert.Error(t, err)
-	assert.Equal(t, err.Error(), "used anchor address is not the node configured address")
+	assert.Equal(t, err.Error(), "anchor address is not the node configured address")
```
Comment (vimukthi-git) [inline documents/validator_test.go]: why don't you use your own function ? :) 
```
assert.True(t, errors.IsOfType(ErrDifferentAnchoredAddress, err))
```

**Comment by ved** [comment]
Vimukti,
My dear, I really don't know why am I into this conversation. Please remove
me

**Comment by vimukthi-git** [comment]
@ved I didn't add you, please unsubscribe. I only replied when you first commented
```
What is it all about?
```
I think this was some github glitch

**Pull request merged**